app/src/crypto_helper.c

```diff
@@ -371,24 +363,68 @@ parser_error_t computeMasterFromSeed(const uint8_t seed[KEY_LENGTH],  uint8_t ma
     return parser_ok;
 }
 
+bool check_diversifier(const uint8_t d[DIVERSIFIER_LENGTH]) {
+    if(d == NULL) {
+        return parser_unexpected_error;
+    }
+
+    uint8_t hash[32] = {0};
+
+    blake2s_state state = {0};
+    blake2s_init_with_personalization(&state, 32, (const uint8_t *)KEY_DIVERSIFICATION_PERSONALIZATION, sizeof(KEY_DIVERSIFICATION_PERSONALIZATION));
+    blake2s_update(&state, (const uint8_t *)GH_FIRST_BLOCK, sizeof(GH_FIRST_BLOCK));
+    blake2s_update(&state, d, DIVERSIFIER_LENGTH);
+    blake2s_final(&state, hash, KEY_LENGTH);
+
+    return is_valid_diversifier(hash);
+}
+
 parser_error_t computeDiversifiersList(const uint8_t dk[KEY_LENGTH], uint8_t div_start_index[11], uint8_t div_l[44]) {
     if(dk == NULL || div_start_index == NULL || div_l == NULL) {
         return parser_unexpected_error;
     }
 
    return get_default_diversifier_list(dk, div_start_index, div_l);
 }
+
 parser_error_t computeDiversifier(const uint8_t dk[KEY_LENGTH], uint8_t div_start_index[11], uint8_t div[11]) {
-    if(dk == NULL || div_start_index == NULL || div == NULL) {
-        return parser_unexpected_error;
+    bool found = false;
+    uint8_t d_list[4*DIVERSIFIER_LENGTH] = {0};
+
+    while (!found)
+    {
+        CHECK_ERROR(computeDiversifiersList(dk, div_start_index, d_list));
```

Comment:
Please, add some comments about how this is working or the expected result from each part. For instance, comment that `computeDiversifiersList` returns a list of 4 diversifiers. Also check that this might end up with an infinitive loop. If none of the 4 diversifiers worked, then it will be blocked in the while loop.